### PR TITLE
[BUGFIX] Set l18n_parent to localisation parent

### PR DESCRIPTION
**Pull request by BenjaminBeck**
I fixed the l18n_parent assigment and also expanded the creation of the copyRecord parameters because i think that way it is more easy to understand

### CODE REVIEW

File: Classes/Backend/TceMain.php [hunk 1]
```diff
@@ -80,6 +80,91 @@ public function processDatamap_postProcessFieldArray($status, $table, $id, &$fie
...
+        $doProcessCommand = in_array($command,['delete']);
+        $doProcessCommand = $doProcessCommand && 'tt_content' == $table;
+
+
+        if ($doProcessCommand) {
+            // A copy, or localisation (which is also a copy) was made. Cascade copy operations for child records.
+            $gridChildRecords = [];
+            $resolver = $this->objectManager->get(ProviderResolver::class);
+            $originalRecord = $this->recordService->getSingle($table, '*', $id);
+            $recordCanBeGrid = !empty($originalRecord['pi_flexform']);
```
Comment (NamelessCoder) [inline Classes/Backend/TceMain.php]: This assumption is wrong - any record can be a grid, regardless whether it uses a flexform.

File: Classes/Backend/TceMain.php [hunk 2]
```diff
@@ -80,6 +80,91 @@ public function processDatamap_postProcessFieldArray($status, $table, $id, &$fie
...
+                $originalRecord
+            );
+            $colPosCalcBaseUid = $id;
+            $gridIsTranslation = $originalRecord['sys_language_uid'] > 0;
+            if($gridIsTranslation){
+                $colPosCalcBaseUid = $originalRecord['l18n_parent'];
+            }
+            if ($primaryProvider) {
+                $childColPosValues = [];
+                foreach ($primaryProvider->getGrid($originalRecord)->getRows() as $row) {
```
Comment (NamelessCoder) [inline Classes/Backend/TceMain.php]: I think this now makes sense as a method on Grid, like `buildColumnPositionValuesForParent(int $parentUid)` which returns a plain array of all possible colPos values in a grid.

File: Classes/Backend/TceMain.php [hunk 3]
```diff
@@ -80,6 +80,91 @@ public function processDatamap_postProcessFieldArray($status, $table, $id, &$fie
...
+            $gridHasChildren = count($gridChildRecords) > 0;
+
+            if('delete' == $command && $gridHasChildren){
+                $cmd = [
+                    'tt_content' => []
+                ];
+                foreach ($gridChildRecords as $recordToDelete) {
+                    $cmd['tt_content'][ $recordToDelete['uid'] ] = ['delete' => 1];
+                }
+                $dataHandler = GeneralUtility::makeInstance(DataHandler::class);
```
Comment (NamelessCoder) [inline Classes/Backend/TceMain.php]: Call on $reference instead - and I think this needs to be different. This is the command pre-processing method so we can simply extend the command map array on $reference and it should then do those, between calling this method and the post-processing method.

File: Classes/Backend/TceMain.php [hunk 4]
```diff
@@ -102,11 +187,18 @@ public function processDatamap_postProcessFieldArray($status, $table, $id, &$fie
...
+        $doProcessCommand = in_array($command,['localize','copy','copyToLanguage','delete']);
+        $doProcessCommand = $doProcessCommand && 'tt_content' == $table;
+
+        if ($doProcessCommand) {
             // A copy, or localisation (which is also a copy) was made. Cascade copy operations for child records.
-            $recordsToCopy = [];
+            $gridChildRecords = [];
             $resolver = $this->objectManager->get(ProviderResolver::class);
             $originalRecord = $this->recordService->getSingle($table, '*', $id);
+            $recordCanBeGrid = !empty($originalRecord['pi_flexform']);
```
Comment (NamelessCoder) [inline Classes/Backend/TceMain.php]: Looks like we have significant duplication here - maybe there's a way we can refactor!

**Comment by NamelessCoder** [comment]
I've taken the liberty of refactoring and recreating the operations you had added here - so although this PR won't get merged, I give you full credit for pointing out the need for improvement ;)

**Comment by BenjaminBeck** [comment]
Ok, i will then push the fixes and changes i am currently working on into my branch and we will then figure out how to merge

**Comment by NamelessCoder** [comment]
You should pull from `feature/colpos` first, I've changed TceMain, I picked the essential stuff from this PR and refactored the original PR - those changes I hope solve the same as your changes in this branch. I'm not sure we need further changes in TceMain...

**Pull request closed without merging**